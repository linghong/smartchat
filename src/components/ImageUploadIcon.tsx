import React, { useRef, ChangeEvent, FC } from 'react'
import { BsImageFill  } from "react-icons/bs"
interface ImageUploadIconProps {
  onImageUpload: (file: File) => void
}

const ImageUploadIcon: FC<ImageUploadIconProps> = ({ onImageUpload }) => {
  const fileInputRef = useRef<HTMLInputElement>(null)

  // Trigger the file input when onKeyDown is fired
  const handleIconClick = () => {
    fileInputRef.current?.click()
  };

  const handleFileChange = (event: ChangeEvent<HTMLInputElement>) => {
    const file = event.target.files ? event.target.files[0] : null 
    if (file) {
      onImageUpload(file)
    }
  }

  return ( 
    <label 
      className="flex items-center justify-center  text-white font-bold px-2 rounded cursor-pointer"
      tabIndex={0}
      onKeyDown={(e) => e.key === 'Enter' && handleIconClick()}
    >
      <input 
        type="file" 
        ref={fileInputRef}
        onChange={handleFileChange} 
        style={{ display: 'none' }} 
        accept="image/*"
        aria-label="Upload image"       
      />
      <BsImageFill 
        size={24} 
        style={{ cursor: 'pointer', background: '#7b7bdc', border:'solid black 1px' }}
        aria-hidden="true" // Hides the icon from screen readers
      />
    </label>
  )
}

export default ImageUploadIcon
