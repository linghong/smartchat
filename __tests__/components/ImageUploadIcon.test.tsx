import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import ImageUploadIcon from '@/src/components/ImageUploadIcon'

describe('ImageUploadIcon', () => {
  const onImageUpload = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should render the upload icon', () => {
    render(<ImageUploadIcon onImageUpload={onImageUpload} />)

    const icon = screen.getByLabelText('Upload image')
    expect(icon).toBeInTheDocument()
  })

  it('should trigger file input click when icon is clicked', () => {
    render(<ImageUploadIcon onImageUpload={onImageUpload} />)

    const icon = screen.getByLabelText('Upload image')
    fireEvent.click(icon)

    const fileInput = screen.getByLabelText('Upload image')
    expect(fileInput).toBeInTheDocument()
    expect(fileInput).toHaveProperty('type', 'file')
  })

  it('should trigger file input click when Enter key is pressed', () => {
    render(<ImageUploadIcon onImageUpload={onImageUpload} />)

    const icon = screen.getByLabelText('Upload image')
    fireEvent.keyDown(icon, { key: 'Enter', code: 'Enter' })

    const fileInput = screen.getByLabelText('Upload image')
    expect(fileInput).toBeInTheDocument()
    expect(fileInput).toHaveProperty('type', 'file')
  })

  it('should call onImageUpload when a file is selected', () => {
    render(<ImageUploadIcon onImageUpload={onImageUpload} />)

    const fileInput = screen.getByLabelText('Upload image') as HTMLInputElement
    const file = new File(['dummy content'], 'example.png', {
      type: 'image/png'
    })
    fireEvent.change(fileInput, { target: { files: [file] } })

    expect(onImageUpload).toHaveBeenCalledTimes(1)
    expect(onImageUpload).toHaveBeenCalledWith(file)
  })

  it('should match snapshot', () => {
    const { asFragment } = render(
      <ImageUploadIcon onImageUpload={onImageUpload} />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
