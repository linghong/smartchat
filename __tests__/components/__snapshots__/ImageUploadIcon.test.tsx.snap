// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUploadIcon should match snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex items-center justify-center  text-white font-bold px-2 rounded cursor-pointer"
    tabindex="0"
  >
    <input
      accept="image/*"
      aria-label="Upload image"
      style="display: none;"
      type="file"
    />
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      style="cursor: pointer; background: rgb(123, 123, 220); border: 1px solid black;"
      viewBox="0 0 16 16"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M.002 3a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-12a2 2 0 0 1-2-2zm1 9v1a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V9.5l-3.777-1.947a.5.5 0 0 0-.577.093l-3.71 3.71-2.66-1.772a.5.5 0 0 0-.63.062zm5-6.5a1.5 1.5 0 1 0-3 0 1.5 1.5 0 0 0 3 0"
      />
    </svg>
  </label>
</DocumentFragment>
`;
